packages/babel-parser/src/plugins/placeholders.js

```diff
@@ -0,0 +1,313 @@
+      // expressions and binding patterns, since they are already
+      // handled by the parseExprAtom and parseBindingAtom functions.
+      // This is needed, for example, to parse "class %%NAME%% {}".
+      return (
+        this.parsePlaceholder("Identifier") ||
+        super.parseIdentifier(...arguments)
+      );
+    }
+
+    checkReservedWord(word: string): void {
+      // Sometimes we call #checkReservedWord(node.name), expecting
+      // that node is an Identifier. If it is a Placeholder, name
+      // will be undefined.
+      if (word !== undefined) super.checkReservedWord(...arguments);
+    }
+
+    /* ============================================================ *
+     * parser/lval.js                                               *
+     * ============================================================ */
+
+    parseBindingAtom(): MaybePlaceholder<"Pattern"> {
+      return (
+        this.parsePlaceholder("Pattern") || super.parseBindingAtom(...arguments)
+      );
+    }
+
+    checkLVal(expr: N.Expression): void {
+      if (expr.type !== "Placeholder") super.checkLVal(...arguments);
+    }
+
+    toAssignable(node: N.Node): N.Node {
+      if (
+        node &&
+        node.type === "Placeholder" &&
+        node.expectedNode === "Expression"
+      ) {
+        node.expectedNode = "Pattern";
+        return node;
+      }
+      return super.toAssignable(...arguments);
+    }
+
+    /* ============================================================ *
+     * parser/statement.js                                          *
+     * ============================================================ */
+
+    verifyBreakContinue(node: N.BreakStatement | N.ContinueStatement) {
+      if (node.label && node.label.type === "Placeholder") return;
+      super.verifyBreakContinue(...arguments);
+    }
+
+    parseExpressionStatement(
+      node: MaybePlaceholder<"Statement">,
+      expr: N.Expression,
+    ): MaybePlaceholder<"Statement"> {
+      if (
+        expr.type !== "Placeholder" ||
+        (expr.extra && expr.extra.parenthesized)
+      ) {
+        return super.parseExpressionStatement(...arguments);
+      }
+
+      if (this.match(tt.colon)) {
+        const stmt: N.LabeledStatement = node;
+        stmt.label = this.finishPlaceholder(expr, "Identifier");
+        this.next();
+        stmt.body = this.parseStatement("label");
+        return this.finishNode(stmt, "LabeledStatement");
+      }
+
+      this.semicolon();
+
+      node.name = expr.name;
+      return this.finishPlaceholder(node, "Statement");
+    }
+
+    parseBlock(): MaybePlaceholder<"BlockStatement"> {
+      return (
+        this.parsePlaceholder("BlockStatement") ||
+        super.parseBlock(...arguments)
+      );
+    }
+
+    parseFunctionId(): ?MaybePlaceholder<"Identifier"> {
+      return (
+        this.parsePlaceholder("Identifier") ||
+        super.parseFunctionId(...arguments)
+      );
+    }
```

Comment:
Is this function even necessary if we overwrite `parseIdentifier()` anyway ?

Reply:
Yes, because `this.match(tt.name)` is false for placeholders and thus it isn't parsed.

Reply:
Ah true.